dask_planner/src/sql/optimizer/dynamic_partition_pruning.rs

```diff
@@ -0,0 +1,968 @@
+//! filters in a table scan. This allows us to eliminate all other rows which do not fit the join
+//! condition from being read at all.
+//!
+//! Furthermore, a table involved in a join may be filtered during a scan, which allows us to
+//! further prune the values to be read.
+
+use std::{
+    collections::{HashMap, HashSet},
+    fs,
+    hash::{Hash, Hasher},
+};
+
+use datafusion_python::{
+    datafusion::parquet::{
+        basic::Type as BasicType,
+        file::reader::{FileReader, SerializedFileReader},
+        record::{reader::RowIter, RowAccessor},
+        schema::{parser::parse_message_type, types::Type},
+    },
+    datafusion_common::{Column, Result, ScalarValue},
+    datafusion_expr::{
+        logical_plan::LogicalPlan,
+        utils::from_plan,
+        Expr,
+        JoinType,
+        Operator,
+        TableScan,
+    },
+    datafusion_optimizer::{OptimizerConfig, OptimizerRule},
+};
+use log::warn;
+
+use crate::sql::table::DaskTableSource;
+
+// Optimizer rule for dynamic partition pruning
+pub struct DynamicPartitionPruning {}
+
+impl DynamicPartitionPruning {
+    #[allow(missing_docs)]
+    #[allow(dead_code)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for DynamicPartitionPruning {
+    fn name(&self) -> &str {
+        "dynamic_partition_pruning"
+    }
+
+    fn try_optimize(
+        &self,
+        plan: &LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        // Parse the LogicalPlan and store tables and columns being (inner) joined upon. We do this
+        // by creating a HashSet of all InnerJoins' join.on and join.filters
+        let join_conds = gather_joins(plan);
+        let tables = gather_tables(plan);
+        let aliases = gather_aliases(plan);
+
+        if join_conds.is_empty() || tables.is_empty() {
+            // No InnerJoins to optimize with
+            Ok(None)
+        } else {
+            // Find the size of the largest table in the query
+            let mut largest_size = 1_f64;
+            for table in &tables {
+                let table_size = table.1.size.unwrap_or(0) as f64;
+                if table_size > largest_size {
+                    largest_size = table_size;
+                }
+            }
+
+            let mut join_values = vec![];
+            let mut join_tables = vec![];
+            let mut join_fields = vec![];
+            let mut fact_tables = HashSet::new();
+
+            // Iterate through all inner joins in the query
+            for join_cond in &join_conds {
+                let join_on = &join_cond.on;
+                for on_i in join_on {
+                    // Obtain tables and columns (fields) involved in join
+                    let left_on = &on_i.0;
+                    let right_on = &on_i.1;
+                    let mut left_table: Option<String> = None;
+                    let mut left_field: Option<String> = None;
```

Comment:
Since this is used as a "pair" might make sense to create a tuple like `(Option<String>, Option<String>)` to keep the variables stored "together"

Reply:
Yeah, I like this idea, thanks!